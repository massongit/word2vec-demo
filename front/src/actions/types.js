/**
 * 分散表現による計算結果の表示Actionを表すタイプ
 * @type {string}
 */
export const SHOW_SIMILAR_WORDS = "SHOW_SIMILAR_WORDS"

/**
 * ローディングActionを表すタイプ
 * @type {string}
 */
export const LOADING = "LOADING"

/**
 * 手法一覧セットActionを表すタイプ
 * @type {string}
 */
export const SET_METHODS = "SET_METHODS"

/**
 * 手法一覧Actionを表すタイプ
 * @type {string}
 */
export const SET_METHOD = "SET_METHOD"
