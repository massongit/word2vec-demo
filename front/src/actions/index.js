import * as types from "./types"
import {createAction} from "redux-actions"

/**
 * Word2Vecによる計算結果の表示ActionのActionCreator
 * @type {function}
 */
export const showSimilarWords = createAction(types.SHOW_SIMILAR_WORDS, parameter => ({
    words: parameter.words,
    keywords: parameter.keywords
}))
