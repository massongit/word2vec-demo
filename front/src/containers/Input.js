import Input from "../components/Input"
import {connect} from "react-redux"
import {injectIntl} from "react-intl"

const mapStateToProps = state => ({
    keywords: state.showSimilarWords.keywords,
    loading: state.loading.loading,
    prevMethod: state.showSimilarWords.method,
    method: state.setMethod.method
})

/**
 * 入力部のContainer
 */
export default injectIntl(connect(mapStateToProps)(Input))
