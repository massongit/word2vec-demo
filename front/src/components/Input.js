import React from "react"
import fetch from "node-fetch"
import PropTypes from "prop-types"
import {Button, Form, FormControl, InputGroup} from "react-bootstrap"
import {FormattedMessage, intlShape} from "react-intl"
import {showSimilarWords} from "../actions"
import {propTypesPN} from "./KeyWords"

/**
 * 入力部
 */
class Input extends React.Component {
    static propTypes = {
        dispatch: PropTypes.func.isRequired,
        keywords: PropTypes.shape(propTypesPN).isRequired,
        intl: intlShape.isRequired
    }

    constructor(props) {
        super(props)
        this.onSubmit = this.onSubmit.bind(this)
    }

    /**
     * 入力されたキーワードをpositiveな単語とnegativeな単語に分類する
     * @returns {{positive: Array, negative: Array}} APIへの入力用のキーワード
     */
    classifyKeyWords() {
        // キーワード
        const keywords = {
            positive: [],
            negative: []
        }

        for (const word of this.input.value.replace("　", " ").trim().split(" ")) {
            if (word.startsWith("-")) {
                keywords.negative.push(word.slice(1))
            } else if (word.startsWith("+")) {
                keywords.positive.push(word.slice(1))
            } else if (word) {
                keywords.positive.push(word)
            }
        }

        return keywords
    }

    /**
     * 入力されたキーワードを元にAPIへの入力用のキーワードを作成する
     * @returns {{positive: Array, negative: Array}} APIへの入力用のキーワード
     */
    makeKeyWords() {
        // キーワード
        const keywords = this.classifyKeyWords()

        // キーワードのりストをソート
        for (const k in keywords) {
            keywords[k].sort()
        }

        return keywords
    }

    /**
     * onSubmitイベント
     * @param ev イベント
     */
    async onSubmit(ev) {
        // サーバーへのSubmitが行われないようにする
        ev.preventDefault()

        // キーワード
        const keywords = this.makeKeyWords()

        // 以前の入力文とは異なる文章が入力されたとき、Word2Vecによる計算を行い、Word2Vecによる計算結果の表示Actionをdispatch
        if ((0 < keywords.positive.length || 0 < keywords.negative.length) && JSON.stringify(keywords) !== JSON.stringify(this.props.keywords)) {
            try {
                this.props.dispatch(showSimilarWords({
                    keywords,
                    words: await this.callWord2Vec(keywords)
                }))
            } catch (er) {
                alert(this.props.intl.formatMessage(
                    {
                        id: "errorMessage.predict"
                    },
                    {
                        message: er.message
                    }
                ))
            }
        }
    }

    /**
     * Word2VecのAPIを呼び出す
     * @param query {object} クエリ
     * @returns {Promise<*>} 計算結果
     */
    async callWord2Vec(query) {
        return await (await fetch(
            "/word2vec",
            {
                method: "POST",
                headers: {
                    "Accept": "application/json",
                    "Content-Type": "application/json"
                },
                body: JSON.stringify(query)
            }
        ).then(res => {
            if (res.ok) {
                return res
            } else {
                throw new Error(res.statusText)
            }
        })).json()
    }

    render() {
        return (
            <Form
                onSubmit={this.onSubmit}
            >
                <InputGroup>
                    <FormControl
                        defaultValue="王 -男 女"
                        inputRef={n => {
                            this.input = n
                        }}
                    />
                    <InputGroup.Button>
                        <Button type="submit">
                            <FormattedMessage id="calculate"/>
                        </Button>
                    </InputGroup.Button>
                </InputGroup>
            </Form>
        )
    }
}

export default Input
