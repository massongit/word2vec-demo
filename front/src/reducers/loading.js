import * as types from "../actions/types"
import {handleActions} from "redux-actions"

/**
 * ローディングActionのReducer
 */
export default handleActions(
    {
        // ローディングActionのとき
        [types.LOADING]: (state, action) => {
            if (typeof action.payload.loading === "boolean") {
                return {
                    loading: action.payload.loading
                }
            } else {
                return state
            }
        }
    },
    {
        loading: false
    }
)
