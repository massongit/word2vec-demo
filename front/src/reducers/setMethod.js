import * as types from "../actions/types"
import {handleActions} from "redux-actions"

/**
 * 手法セットActionのReducer
 */
export default handleActions(
    {
        // 手法セットActionのとき
        [types.SET_METHOD]: (state, action) => {
            if (action.payload.method) {
                return {
                    prevMethod: state.method,
                    method: action.payload.method
                }
            } else {
                return state
            }
        }
    }, {}
)
