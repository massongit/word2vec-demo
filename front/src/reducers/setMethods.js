import * as types from "../actions/types"
import {handleActions} from "redux-actions"

/**
 * 手法一覧セットActionのReducer
 */
export default handleActions(
    {
        // 手法一覧セットActionのとき
        [types.SET_METHODS]: (state, action) => {
            if (0 < action.payload.methods.length) {
                return {
                    methods: action.payload.methods
                }
            } else {
                return state
            }
        }
    },
    {
        methods: []
    }
)
