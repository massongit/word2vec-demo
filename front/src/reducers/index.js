import loading from "./loading"
import setMethod from "./setMethod"
import setMethods from "./setMethods"
import showSimilarWords from "./showSimilarWords"
import {combineReducers} from "redux"

/**
 * ルートReducer
 */
export default combineReducers({
    showSimilarWords,
    loading,
    setMethods,
    setMethod
})
