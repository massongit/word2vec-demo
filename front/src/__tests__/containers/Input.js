import React from "react"
import Input from "../../containers/Input"
import thunk from "redux-thunk"
import fetch from "node-fetch"
import configureMockStore from "redux-mock-store"
import words2 from "../../test_data/words2"
import words from "../../test_data/words"
import initialLoadingState from "../../test_data/initialLoadingState"
import initialShowSimilarWordsState from "../../test_data/initialShowSimilarWordsState"
import * as types from "../../actions/types"
import {Form, FormControl} from "react-bootstrap"
import {loadTranslation, mountWithIntl, shallowWithIntl} from "enzyme-react-intl"
import {showSimilarWordsState, showSimilarWordsState2} from "../../test_data"
import {makeShowSimilarWordsAction} from "../reducers"

export const eventMock = {
    preventDefault: jest.fn()
}

const rootStateAfterShowSimilarWords = {
    loading: initialLoadingState,
    showSimilarWords: showSimilarWordsState
}

const checkContainShowSimilarWords = (store, inputComponent, w) => {
    let isExpect = false
    const action_ = makeShowSimilarWordsAction(w)
    for (const action of store.getActions()) {
        if (action.type === types.SHOW_SIMILAR_WORDS) {
            expect(action).toEqual(action_)
            isExpect = true
        }
    }
    expect(isExpect).toBeTruthy()
}

const keywords_input = "-男 王 女"

loadTranslation("./src/translations/ja.json")

describe("containers/Input", () => {
    beforeEach(() => {
        fetch.resetMocks()
    })

    it("Componentが正しく配置されている", () => {
        const store = configureMockStore([thunk])({
            loading: initialLoadingState,
            showSimilarWords: initialShowSimilarWordsState
        })
        const inputComponent = shallowWithIntl(
            <Input
                store={store}
            />
        ).dive()
        expect(inputComponent).toMatchSnapshot()
    })

    it("Formになっている", () => {
        const store = configureMockStore([thunk])({
            loading: initialLoadingState,
            showSimilarWords: initialShowSimilarWordsState
        })
        const inputComponent = mountWithIntl(
            <Input
                store={store}
            />
        )
        expect(inputComponent.contains(Form)).toBeTruthy()
    })

    it("onSubmitイベントが呼び出されたとき、サーバーへのSubmitが行われない", async () => {
        const store = configureMockStore([thunk])({
            loading: initialLoadingState,
            showSimilarWords: initialShowSimilarWordsState
        })
        const inputComponent = mountWithIntl(
            <Input
                store={store}
            />
        )
        await inputComponent.find(Form).props().onSubmit(eventMock)
        expect(eventMock.preventDefault.mock.calls).toHaveLength(1)
    })

    it("入力文が空の状態で、onSubmitイベントが呼び出されたとき、fetchとshowSimilarWordsのdispatchが行われない", async () => {
        const store = configureMockStore([thunk])({
            loading: initialLoadingState,
            showSimilarWords: initialShowSimilarWordsState
        })
        const inputComponent = mountWithIntl(
            <Input
                store={store}
            />
        )
        inputComponent.find(FormControl).children().instance().value = ""
        await inputComponent.find(Form).props().onSubmit(eventMock)
        expect(fetch.mock.calls).toHaveLength(0)
        for (const action of store.getActions()) {
            expect(action.type).not.toEqual([types.SHOW_SIMILAR_WORDS])
        }
    })

    it("入力文が入力された状態で、onSubmitイベントが呼び出されたとき、fetchとshowSimilarWordsのdispatchが正常に行われる", async () => {
        const store = configureMockStore([thunk])({
            loading: initialLoadingState,
            showSimilarWords: initialShowSimilarWordsState
        })
        const inputComponent = mountWithIntl(
            <Input
                store={store}
            />
        )
        inputComponent.find(FormControl).children().instance().value = keywords_input
        fetch.mockResponse(JSON.stringify(words))
        await inputComponent.find(Form).props().onSubmit(eventMock)
        expect(fetch.mock.calls).toHaveLength(1)
        checkContainShowSimilarWords(store, inputComponent, showSimilarWordsState)
    })

    it("前回と同じ入力内容でonSubmitイベントを呼び出したとき、fetchやshowSimilarWordsのdispatchが行われない", async () => {
        const store = configureMockStore([thunk])(rootStateAfterShowSimilarWords)
        const inputComponent = mountWithIntl(
            <Input
                store={store}
            />
        )
        inputComponent.find(FormControl).children().instance().value = keywords_input
        await inputComponent.find(Form).props().onSubmit(eventMock)
        expect(fetch.mock.calls).toHaveLength(0)
        for (const action of store.getActions()) {
            expect(action.type).not.toEqual(types.SHOW_SIMILAR_WORDS)
        }
    })

    it("前回とは違う入力内容でonSubmitイベントを呼び出したとき、fetchやshowSimilarWordsのdispatchが正常に行われる", async () => {
        const store = configureMockStore([thunk])(rootStateAfterShowSimilarWords)
        const inputComponent = mountWithIntl(
            <Input
                store={store}
            />
        )
        inputComponent.find(FormControl).children().instance().value = "-日本 東京 アメリカ"
        fetch.mockResponse(JSON.stringify(words2))
        await inputComponent.find(Form).props().onSubmit(eventMock)
        expect(fetch.mock.calls).toHaveLength(1)
        checkContainShowSimilarWords(store, inputComponent, showSimilarWordsState2)
    })
})
