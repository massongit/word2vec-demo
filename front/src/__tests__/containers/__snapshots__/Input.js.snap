// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Input Componentが正しく配置されている 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(Input)
    intl={
        Object {
            "defaultFormats": Object {},
            "defaultLocale": "en",
            "formatDate": [Function],
            "formatHTMLMessage": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatPlural": [Function],
            "formatRelative": [Function],
            "formatTime": [Function],
            "formats": Object {},
            "formatters": Object {
              "getDateTimeFormat": [Function],
              "getMessageFormat": [Function],
              "getNumberFormat": [Function],
              "getPluralFormat": [Function],
              "getRelativeFormat": [Function],
            },
            "locale": "en",
            "messages": Object {
              "calculate": "計算",
              "description.input": "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
              "description.input.supplement.1": "引き算を行う場合には、単語の先頭に『-』を付けます。",
              "description.input.supplement.2": "足し算を行う場合には、単語の先頭に『+』を付けることができます (基本的には付けなくても問題ありませんが、単語が『+』や『-』で始まる場合には必ず付けてください)。",
              "description.word2Vec": "単語をベクトル表現する技術です。これを利用して、単語の足し算や引き算を行うことができます。",
              "errorMessage.default": "エラーが発生しました: {message}",
              "errorMessage.predict": "計算中にエラーが発生しました: {message}",
              "header.candidate": "候補",
              "header.cosine": "コサイン類似度 (-1 〜 1)",
              "header.keywords": "<キーワード>",
              "header.word": "単語",
              "title.description": "概要",
              "title.input": "入力",
              "title.result": "結果",
              "unknown": "(予測不能)",
            },
            "now": [Function],
            "textComponent": "span",
          }
    }
    store={
        Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "dispatch": [Function],
      "intl": Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "calculate": "計算",
          "description.input": "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
          "description.input.supplement.1": "引き算を行う場合には、単語の先頭に『-』を付けます。",
          "description.input.supplement.2": "足し算を行う場合には、単語の先頭に『+』を付けることができます (基本的には付けなくても問題ありませんが、単語が『+』や『-』で始まる場合には必ず付けてください)。",
          "description.word2Vec": "単語をベクトル表現する技術です。これを利用して、単語の足し算や引き算を行うことができます。",
          "errorMessage.default": "エラーが発生しました: {message}",
          "errorMessage.predict": "計算中にエラーが発生しました: {message}",
          "header.candidate": "候補",
          "header.cosine": "コサイン類似度 (-1 〜 1)",
          "header.keywords": "<キーワード>",
          "header.word": "単語",
          "title.description": "概要",
          "title.input": "入力",
          "title.result": "結果",
          "unknown": "(予測不能)",
        },
        "now": [Function],
        "textComponent": "span",
      },
      "keywords": Object {
        "negative": Array [],
        "positive": Array [],
      },
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "intl": Object {
          "defaultFormats": Object {},
          "defaultLocale": "en",
          "formatDate": [Function],
          "formatHTMLMessage": [Function],
          "formatMessage": [Function],
          "formatNumber": [Function],
          "formatPlural": [Function],
          "formatRelative": [Function],
          "formatTime": [Function],
          "formats": Object {},
          "formatters": Object {
            "getDateTimeFormat": [Function],
            "getMessageFormat": [Function],
            "getNumberFormat": [Function],
            "getPluralFormat": [Function],
            "getRelativeFormat": [Function],
          },
          "locale": "en",
          "messages": Object {
            "calculate": "計算",
            "description.input": "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
            "description.input.supplement.1": "引き算を行う場合には、単語の先頭に『-』を付けます。",
            "description.input.supplement.2": "足し算を行う場合には、単語の先頭に『+』を付けることができます (基本的には付けなくても問題ありませんが、単語が『+』や『-』で始まる場合には必ず付けてください)。",
            "description.word2Vec": "単語をベクトル表現する技術です。これを利用して、単語の足し算や引き算を行うことができます。",
            "errorMessage.default": "エラーが発生しました: {message}",
            "errorMessage.predict": "計算中にエラーが発生しました: {message}",
            "header.candidate": "候補",
            "header.cosine": "コサイン類似度 (-1 〜 1)",
            "header.keywords": "<キーワード>",
            "header.word": "単語",
            "title.description": "概要",
            "title.input": "入力",
            "title.result": "結果",
            "unknown": "(予測不能)",
          },
          "now": [Function],
          "textComponent": "span",
        },
        "keywords": Object {
          "negative": Array [],
          "positive": Array [],
        },
        "store": Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [Function],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "context": Object {
      "intl": Object {
        "defaultFormats": Object {},
        "defaultLocale": "en",
        "formatDate": [Function],
        "formatHTMLMessage": [Function],
        "formatMessage": [Function],
        "formatNumber": [Function],
        "formatPlural": [Function],
        "formatRelative": [Function],
        "formatTime": [Function],
        "formats": Object {},
        "formatters": Object {
          "getDateTimeFormat": [Function],
          "getMessageFormat": [Function],
          "getNumberFormat": [Function],
          "getPluralFormat": [Function],
          "getRelativeFormat": [Function],
        },
        "locale": "en",
        "messages": Object {
          "calculate": "計算",
          "description.input": "日本語の単語を空白で区切って入力し、計算ボタンをクリックしてください。",
          "description.input.supplement.1": "引き算を行う場合には、単語の先頭に『-』を付けます。",
          "description.input.supplement.2": "足し算を行う場合には、単語の先頭に『+』を付けることができます (基本的には付けなくても問題ありませんが、単語が『+』や『-』で始まる場合には必ず付けてください)。",
          "description.word2Vec": "単語をベクトル表現する技術です。これを利用して、単語の足し算や引き算を行うことができます。",
          "errorMessage.default": "エラーが発生しました: {message}",
          "errorMessage.predict": "計算中にエラーが発生しました: {message}",
          "header.candidate": "候補",
          "header.cosine": "コサイン類似度 (-1 〜 1)",
          "header.keywords": "<キーワード>",
          "header.word": "単語",
          "title.description": "概要",
          "title.input": "入力",
          "title.result": "結果",
          "unknown": "(予測不能)",
        },
        "now": [Function],
        "textComponent": "span",
      },
    },
  },
}
`;
